qiskit/circuit/quantumcircuit.py

```diff
@@ -903,7 +968,7 @@ def compose(
                 clbits = self.clbits[: other.num_clbits]
             if front:
                 # Need to keep a reference to the data for use after we've emptied it.
-                old_data = list(dest.data)
+                old_data = dest._new_data(dest.data)
```

Comment:
This feels like having a Rust-space `CircuitData::copy()` method exposed would help matters / efficiency, especially since it wouldn't need to re-intern all the objects but could just increment the refcounts (if we continue doing that).

Reply:
This is now done in 185d7368cdab3d915f76143a9b2e4e3fc18421d0